velox/functions/sparksql/aggregates/tests/AverageAggregationTest.cpp

```diff
@@ -0,0 +1,619 @@
+        {},
+        {fmt::format("{}(c0)", functionName)},
+        "SELECT avg(c0) FROM tmp");
+
+    testAggregations(
+        [&](auto& builder) {
+          builder.values(vectors).project({"c0 % 2 AS c0_mod_2", "c0"});
+        },
+        {"c0_mod_2"},
+        {fmt::format("{}(c0)", functionName)},
+        "SELECT c0 % 2, avg(c0) FROM tmp group by 1");
+  };
+
+  testFunction("spark_avg");
+}
+
+TEST_F(AverageAggregationTest, avgConstNull) {
+  auto testFunction = [this](const std::string& functionName) {
+    // Have at least two row vectors as it triggers different code paths.
+    auto vectors = {
+        makeRowVector({
+            makeNullableFlatVector<int64_t>({0, 1, 2, 0, 1, 2, 0, 1, 2, 0}),
+            makeNullConstant(TypeKind::BIGINT, 10),
+            makeNullConstant(TypeKind::DOUBLE, 10),
+        }),
+        makeRowVector({
+            makeNullableFlatVector<int64_t>({0, 1, 2, 0, 1, 2, 0, 1, 2, 0}),
+            makeNullConstant(TypeKind::BIGINT, 10),
+            makeNullConstant(TypeKind::DOUBLE, 10),
+        }),
+    };
+
+    createDuckDbTable(vectors);
+
+    testAggregations(
+        vectors,
+        {},
+        {fmt::format("{}(c1)", functionName),
+         fmt::format("{}(c2)", functionName)},
+        "SELECT avg(c1), avg(c2) FROM tmp");
+
+    testAggregations(
+        vectors,
+        {"c0"},
+        {fmt::format("{}(c1)", functionName),
+         fmt::format("{}(c2)", functionName)},
+        "SELECT c0, avg(c1), avg(c2) FROM tmp group by c0");
+
+    testAggregations(
+        vectors,
+        {},
+        {fmt::format("{}(c0)", functionName)},
+        "SELECT avg(c0) FROM tmp");
+  };
+
+  testFunction("spark_avg");
+}
+
+TEST_F(AverageAggregationTest, avgNulls) {
+  auto testFunction = [this](const std::string& functionName) {
+    // Have two row vectors a lest as it triggers different code paths.
+    auto vectors = {
+        makeRowVector({
+            makeNullableFlatVector<int64_t>({0, std::nullopt, 2, 0, 1}),
+            makeNullableFlatVector<int64_t>({0, 1, std::nullopt, 3, 4}),
+            makeNullableFlatVector<double>({0.1, 1.2, 2.3, std::nullopt, 4.4}),
+        }),
+        makeRowVector({
+            makeNullableFlatVector<int64_t>({0, std::nullopt, 2, 0, 1}),
+            makeNullableFlatVector<int64_t>({0, 1, std::nullopt, 3, 4}),
+            makeNullableFlatVector<double>({0.1, 1.2, 2.3, std::nullopt, 4.4}),
+        }),
+    };
+
+    createDuckDbTable(vectors);
+
+    testAggregations(
+        vectors,
+        {},
+        {fmt::format("{}(c1)", functionName),
+         fmt::format("{}(c2)", functionName)},
+        "SELECT avg(c1), avg(c2) FROM tmp");
+
+    testAggregations(
+        vectors,
+        {"c0"},
+        {fmt::format("{}(c1)", functionName),
+         fmt::format("{}(c2)", functionName)},
+        "SELECT c0, avg(c1), avg(c2) FROM tmp group by c0");
+  };
+
+  testFunction("spark_avg");
+}
+
+TEST_F(AverageAggregationTest, avgAllNulls) {
```

Comment:
Is this a copy-paste from Presto's test?

Reply:
Not that case. This test is added to ensure the partial result of avg is (0, 0). Except for this one, other tests in this file are originated from Presto's.

Reply:
It would be nice to avoid copy-pasting large chunks of code. Should we keep only Spark-specific tests here?

Reply:
OK, removed the others and keep this one only.